src/shared/copy.c

```
@@ -1158,7 +1158,8 @@ static int fd_copy_directory(
                                          denylist, subvolumes, hardlink_context, child_display_path, progress_path,
                                          progress_bytes, userdata);
 
-                if (IN_SET(r, -EINTR, -ENOSPC)) /* Propagate SIGINT/SIGTERM and ENOSPC up instantly */
+                /* Propagate unsupported, SIGINT/SIGTERM, and ENOSPC up instantly */
+                if (ERRNO_IS_NOT_SUPPORTED(r) || IN_SET(r, -EINTR, -ENOSPC))
```

hmm, is this really a good idea? I'd expect such errors if you create device nodes on vfat or cifs or so, and i think we should go on in that case, and only return failure once we tried to create with all other inodes.

i mean, that's what tar and cp do, not? if you copy an os tree with devnodes and they cannot be created on the target, they continue and fail late, not early. i think we should do the same.

---

For the record, the 'device node on vfat' case seems to produce `EPERM`.

That said, I can kinda see the logic.  For me it just doesn't seem super useful to get a screen full of the same error over and over.  Can you think of another case that you'd expect to see one of the 'not_supported' errnos and we might want to continue anyway?

---

well, it might happen for any kind of inode type the backing fs doesn't support... unix sockets, fifos, symlinks, who knows...

---

no major objections from me if you just want to close this.  I submitted it as a separate PR for a reason ☺️

---

an alternative: I could limit this to strictly ENOTTY.

---

if it's just about not seeing all that output, i'd be fine if you suppress the error once it has been seen. i.e. define another flag COPY_SUPPRESS_NOT_SUPPORTED or so. Then, once we printed the complaint once, just turn on the flag for the rest of the descent

---

I'd prefer not to do that: I want to see the difference between the failure to do the ioctl on the source vs. the destination.

---

so, what about this: let's just make the verity stuff fatal. i.e. whenever we copy verity info, and it fails with some form of not supported, then turn that into a new, recognizable error, for example ESOCKTNOSUPPORT or so. Then, make that one immediately fatal for the whole thing, but the other kinds of not supported leave as is.

that should address things for you, right?

---

I had the same idea.  I'll do a patch soon.